codec-http/src/main/java/io/netty/handler/codec/http/compression/CompressionOptions.java

```diff
@@ -0,0 +1,27 @@
+/*
+ * Copyright 2021 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.handler.codec.http.compression;
+
+/**
+ * {@link CompressionOptions} provides compression options for
+ * various types of compressor types, like Brotli.
+ *
+ * A {@link CompressionOptions} instance is thread-safe
+ * and should be shared between multiple instances of Compressor.
+ */
+public interface CompressionOptions {
+    // Empty
```

Comment:
Could we add some factory methods here for users' convenience?

```java
public interface CompressionOptions {
    static BrotliCompressionOptions brotli(...) { ... }
    static GzipCompressionOptions gzip(...) { ... }
    static DeflateCompressionOptions deflate(...) { ... }
}
```

Then, we could hide the public constructors of the options classes as well.

Reply:
Not sure about it. Can you write a patch?

Reply:
It's a little too long to write a patch but it basically looks like this:

```java
public interface CompressionOptions {
    static BrotliCompressionOptions brotli(Encoder.Parameters params) {
        return new BrotliCompressionOptions(params);
    }
    static GzipCompressionOptions gzip(int compressionLevel, ...) {
        return new GzipCompressionOptions(compressionLevel, ...);
    }
    static DeflateCompressionOptions deflate(int compressionLevel, ...) {
        return new DeflateCompressionOptions(compressionLevel, ...);
    }
}

public final class BrotliCompressionOptions implements CompressionOptions {
    ...
    // Hide from the public API so it can be instantiated only via CompressionOptions.brotli(...)
    BrotliCompressionOptions(Encoder.Parameters params) {
        this.params = params;
    }
    ...
}
```

Reply:
Sounds good. I will get back on this once I'm done battling with the preface bug. :D

Reply:
![Screenshot (4)](https://user-images.githubusercontent.com/24762260/123038531-33493b00-d40e-11eb-9b05-b032dc0e90f8.png)
Out of luck.

Reply:
I think we need to have a "utility class here".

Something like:

```java
public final class StandardCompressionOptions {
    private StandardCompressionOptions() { }

   static BrotliCompressionOptions brotli(Encoder.Parameters params) {
        return new BrotliCompressionOptions(params);
    }
    static GzipCompressionOptions gzip(int compressionLevel, ...) {
        return new GzipCompressionOptions(compressionLevel, ...);
    }
    static DeflateCompressionOptions deflate(int compressionLevel, ...) {
        return new DeflateCompressionOptions(compressionLevel, ...);
    }
....
}
```

Reply:
@trustin @normanmaurer PTAL at new code.